src/login/logind.c

```diff
@@ -409,10 +409,64 @@ static int manager_enumerate_users(Manager *m) {
         return r;
 }
 
+static int manager_attach_fds(Manager *m) {
+        _cleanup_strv_free_ char **fdnames = NULL;
+        int r = 0, n, i, fd;
+
+        /* Upon restart, PID1 will send us back all fds of session devices
+         * that we previously opened. All file descriptors are associated
+         * with a given session. The session ids are passed through FDNAMES.
+         *
+         * FIXME: should we check that all restored devices have been
+         * reattached with a passed fd ? */
+
+        n = sd_listen_fds_with_names(true, &fdnames);
+        if (n <= 0)
+                return n;
+
+        for (i = 0; i < n; i++) {
+                struct stat st;
+                SessionDevice *sd;
+                Session *s;
+                int k;
+
+                fd = SD_LISTEN_FDS_START + i;
+
+                s = hashmap_get(m->sessions, fdnames[i]);
+                if (!s) {
+                        log_warning("Failed to attach fd for unknown session: %s", fdnames[i]);
+                        r = -EINVAL;
+                        /* FIXME: should we close the fd ? */
+                        continue;
```

Comment:
please be defnesive here. please log about where we can't match up the fdnames list with the session list, but continue. And close all fds we can't find matching sessions for.